File: backend/app/api/v1/events.py
"""
Events API Endpoints
Manages community events (meetings, festivals, maintenance, etc.)
"""
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional
from app.db.session import get_db
from app.api.v1.endpoints.auth import get_current_user
from app.models.user import User, UserRole
from app.models.event import EventType
from app.schemas.event import EventCreate, EventUpdate, EventResponse, EventListResponse
from app.services import event_service


def require_admin(current_user: User = Depends(get_current_user)) -> User:
    """Dependency to require admin role"""
    if current_user.role != UserRole.ADMIN:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Admin access required"
        )
    return current_user

router = APIRouter()


@router.get("/upcoming", response_model=list[EventResponse])
def get_upcoming_events(
    limit: int = Query(5, ge=1, le=20),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get upcoming active events
    
    **Accessible by:** All authenticated users
    """
    events = event_service.get_upcoming_events(db, limit=limit)
    return events


@router.get("", response_model=EventListResponse)
def get_all_events(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=100),
    event_type: Optional[EventType] = None,
    is_active: Optional[bool] = None,
    include_past: bool = True,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get all events with optional filters
    
    **Accessible by:** All authenticated users
    """
    events, total = event_service.get_all_events(
        db,
        skip=skip,
        limit=limit,
        event_type=event_type,
        is_active=is_active,
        include_past=include_past
    )
    return EventListResponse(events=events, total=total)


@router.get("/{event_id}", response_model=EventResponse)
def get_event(
    event_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get event by ID
    
    **Accessible by:** All authenticated users
    """
    event = event_service.get_event_by_id(db, event_id)
    if not event:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Event not found"
        )
    return event


@router.post("", response_model=EventResponse, status_code=status.HTTP_201_CREATED)
def create_event(
    event_data: EventCreate,
    current_user: User = Depends(require_admin),
    db: Session = Depends(get_db)
):
    """
    Create a new event
    
    **Accessible by:** Admin only
    """
    # Validate end_datetime if provided
    if event_data.end_datetime and event_data.end_datetime < event_data.start_datetime:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="End datetime must be after start datetime"
        )
    
    event = event_service.create_event(db, event_data, current_user.id)
    return event


@router.put("/{event_id}", response_model=EventResponse)
def update_event(
    event_id: int,
    event_data: EventUpdate,
    current_user: User = Depends(require_admin),
    db: Session = Depends(get_db)
):
    """
    Update an existing event
    
    **Accessible by:** Admin only
    """
    event = event_service.update_event(db, event_id, event_data)
    if not event:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Event not found"
        )
    return event


@router.delete("/{event_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_event(
    event_id: int,
    hard_delete: bool = Query(False, description="Permanently delete the event"),
    current_user: User = Depends(require_admin),
    db: Session = Depends(get_db)
):
    """
    Delete an event (soft delete by default, hard delete if specified)
    
    **Accessible by:** Admin only
    """
    if hard_delete:
        success = event_service.hard_delete_event(db, event_id)
    else:
        success = event_service.delete_event(db, event_id)
    
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Event not found"
        )
    return None

File: frontend/src/api/events.js
/**
 * Events API Client
 * Handles all API calls related to community events
 */
import api from './client'

export const eventsAPI = {
  /**
   * Get upcoming active events
   * @param {number} limit - Maximum number of events to return (1-20)
   * @returns {Promise<Array>} List of upcoming events
   */
  getUpcoming: (limit = 5) => {
    return api.get(`/events/upcoming?limit=${limit}`)
  },

  /**
   * Get all events with optional filters
   * @param {Object} params - Filter parameters
   * @param {number} params.skip - Number of records to skip
   * @param {number} params.limit - Maximum number of records
   * @param {string} params.event_type - Filter by event type
   * @param {boolean} params.is_active - Filter by active status
   * @param {boolean} params.include_past - Include past events
   * @returns {Promise<Object>} Object with events array and total count
   */
  getAll: (params = {}) => {
    return api.get('/events', { params })
  },

  /**
   * Get event by ID
   * @param {number} id - Event ID
   * @returns {Promise<Object>} Event object
   */
  getById: (id) => {
    return api.get(`/events/${id}`)
  },

  /**
   * Create a new event (Admin only)
   * @param {Object} data - Event data
   * @returns {Promise<Object>} Created event
   */
  create: (data) => {
    return api.post('/events', data)
  },

  /**
   * Update an event (Admin only)
   * @param {number} id - Event ID
   * @param {Object} data - Updated event data
   * @returns {Promise<Object>} Updated event
   */
  update: (id, data) => {
    return api.put(`/events/${id}`, data)
  },

  /**
   * Delete an event (Admin only)
   * @param {number} id - Event ID
   * @returns {Promise<void>}
   */
  delete: (id) => {
    return api.delete(`/events/${id}`)
  }
}

